From 5e82fad410dc1cf72effafeeba475e74f5cc4639 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Wed, 1 Feb 2017 19:01:57 +0200
Subject: [PATCH 0814/1345] fix: net: mvpp2x: fix crashes during mtu change

commit  7be291aff70c920efcc19e641caff017850b9740 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- In case that the Buffer Manager failed to return buffer,
  then 0x0 is returned by BM_PHY_ALLOC and BM_ADDR_HIGH_ALLOC
  didn't cleared. If Buffer Manager failed to return buffer
  BM_ADDR_HIGH_ALLOC should be disregarded.

Change-Id: I968eed0ec76150e8b4aab5b2169d6e440e37af6e
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/36946
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
(cherry picked from commit 1501a6b7cbaee740888c259e97911bafa2c8cc99)
Reviewed-on: http://vgitil04.il.marvell.com:8080/37148
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h
index 790ee39..5ce4d52 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h
@@ -251,6 +251,10 @@ static inline dma_addr_t mv_pp2x_bm_phys_addr_get(struct mv_pp2x_hw *hw, u32 poo
 
 	val = mv_pp2x_read(hw, MVPP2_BM_PHY_ALLOC_REG(pool));
 
+	/*  Disregard BM_ADDR_HIGH_ALLOC if Buffer Manager failed to return buffer */
+	if (!val)
+		return 0;
+
 #ifdef CONFIG_PHYS_ADDR_T_64BIT
 	{
 	u64 val2;
-- 
1.7.9.5

